data_model/derive/src/lib.rs

```diff
@@ -653,3 +478,87 @@ pub fn has_origin_derive(input: TokenStream) -> TokenStream {
 
     emitter.finish_token_stream_with(result)
 }
+
+/// Create an event set structure from an event enum.
+///
+/// Event set is a set of multiple event types, allowing to efficiently test whether an event is in a set.
+///
+/// For this event enum:
+///
+/// ```rust
+/// # use iroha_data_model_derive::EventSet;
+/// #[derive(EventSet)]
+/// pub enum TestEvent {
+///     Event1,
+///     Event2,
+///     NestedEvent(AnotherEvent),
+/// }
+/// pub struct AnotherEvent;
+/// ```
+///
+/// The macro will generate a `TestEventSet` struct.
+///
+/// It will have associated constants that correspond to a singleton set for each event that can be accessed like `TestEventSet::Event1`.
+///
+/// The sets can be combined either with a `|` operator or with the `or` method to match multiple events at once: `TestEventSet::Event1 | TestEventSet::Event2`.
+///
+/// Variants that:
+/// 1. have a single unnamed field
+/// 2. have the type name end with `Event`
+///
+/// are treated as nested events and their set constant has `Any` prepended to the name. For example, `TestEventSet::AnyNestedEvent`.
+///
+/// The set has the following methods:
+/// ```ignore
+/// impl TestEventSet {
+///     /// Returns a set that doesn't contain any events
+///     const fn empty() -> TestEventSet;
+///     /// Returns a set that contains all events
+///     const fn all() -> TestEventSet;
+///     /// Computes union of two sets, const form of the `|` operator
+///     const fn or(self, other: TestEventSet) -> TestEventSet;
+///     /// Checks if the `other` set is a subset of `self` set
+///     const fn contains(&self, other: Self) -> bool;
+///     /// Decomposes an `EventSet` into a vector of basis `EventSet`s, each containing a single event
+///     fn decompose(&self) -> Vec<TestEvent>
```

Comment:
It is used in `Serialize` and I thought it would be useful to anyone wanting to inspect the event filters. It's much easier than testing for each event manually.

I am also a bit conflicted on whether we should expose a separate type for event discriminant: it would provide a nicer API here, but exposing it just for inspecting enums? Not sure if it would be a common thing to do.

Reply:
let's keep this function private for now